packager/media/formats/mp2t/mpeg1_header.cc

```diff
@@ -0,0 +1,228 @@
+  RCHECK(version_ != kMpeg1V_INV);
+  RCHECK(frame.ReadBits(2, &layer_));
+  RCHECK(layer_ != kMpeg1L_INV);
+  RCHECK(frame.ReadBits(1, &protection_absent_));
+
+  uint8_t btr_idx;
+  RCHECK(frame.ReadBits(4, &btr_idx));
+  bitrate_ = Mpeg1BitRate(btr_idx, version_, layer_);
+
+  uint8_t sr_idx;
+  RCHECK(frame.ReadBits(2, &sr_idx));
+  sample_rate_ = Mpeg1SampleRate(sr_idx, version_);
+
+  RCHECK(frame.ReadBits(1, &padded_));
+  // Skip private stream bit.
+  RCHECK(frame.SkipBits(1));
+
+  RCHECK(frame.ReadBits(2, &channel_mode_));
+  // Skip Mode extension
+  RCHECK(frame.SkipBits(2));
+  // Skip copyright, origination and emphasis info.
+  RCHECK(frame.SkipBits(4));
+
+  return true;
+}
+
+size_t Mpeg1Header::GetHeaderSize() const {
+  // Unlike ADTS, for MP3, the whole frame is included in the media sample, so
+  // return 0 header size.
+  return 0;
+}
+
+size_t Mpeg1Header::GetFrameSize() const {
+  return Mpeg1FrameSize(layer_, bitrate_, sample_rate_, padded_);
+}
+
+size_t Mpeg1Header::GetFrameSizeWithoutParsing(const uint8_t* data,
+                                               size_t num_bytes) const {
+  DCHECK_GT(num_bytes, static_cast<size_t>(1));
+  uint8_t version = (data[1] & 0b00011000) >> 3;
+  uint8_t layer = (data[1] & 0b00000110) >> 1;
+  uint8_t btr_idx = (data[2] & 0b11110000) >> 4;
+  uint8_t sr_idx = (data[2] & 0b00001100) >> 2;
+  uint8_t padded = (data[2] & 0b00000010) >> 1;
+
+  int bitrate = Mpeg1BitRate(btr_idx, version, layer);
+  int samplerate = Mpeg1SampleRate(sr_idx, version);
+  return Mpeg1FrameSize(layer, bitrate, samplerate, padded);
+}
+
+void Mpeg1Header::GetAudioSpecificConfig(std::vector<uint8_t>* buffer) const {
+  // The following conversion table is extracted from ISO 14496 Part 3 -
+  // Table 1.16 - Sampling Frequency Index.
+  static const int kConfigFrequencyTable[] = {
+    96000, 88200, 64000, 48000, 44100,
+    32000, 24000, 22050, 16000, 12000,
+    11025, 8000,  7350};
+  static const size_t kConfigFrequencyTableSize = arraysize(kConfigFrequencyTable);
+  uint8_t cft_idx;
+
+  for (cft_idx = 0; cft_idx < kConfigFrequencyTableSize; cft_idx++)
+    if (sample_rate_ == kConfigFrequencyTable[cft_idx])
+      break;
+
+  DCHECK(buffer);
+  buffer->clear();
+  BitWriter config(buffer);
+  config.WriteBits(GetObjectType(), 5);
+  config.WriteBits(cft_idx, 4);
+  /*
+   * NOTE: Number of channels matches channel_configuration index,
+   * since mpeg1 has only 1 or 2 channels
+   */
+  config.WriteBits(GetNumChannels(), 4);
+  config.Flush();
+}
+
+uint8_t Mpeg1Header::GetObjectType() const {
+  /*
+   * ISO14496-3:2009 Table 1.17 - Audio Object Types
+   */
+  if (layer_ == kMpeg1L_1)
+    return 32;
+  else if (layer_ == kMpeg1L_2)
+    return 33;
+
+  RCHECK(layer_ == kMpeg1L_3);
+  return 34;
+}
+
+uint32_t Mpeg1Header::GetSamplingFrequency() const {
+  return sample_rate_;
+}
+
+uint8_t Mpeg1Header::GetNumChannels() const {
+  if (channel_mode_ == 0b11)
+    return 1;
+  return 2;
```

Comment:
So number of channels is 2 if channel_mode_ is 0, 1 or 2?

Reply:
correct! All bits set in channel mode means mono, otherwise 2 channels are used, not necessary to be stereo however.  See https://www.datavoyage.com/mpgscript/mpeghdr.htm